Make ListBuffer's iterator fail when the buffer is mutated

NthPortal opened the pull request:
Make `ListBuffer`'s iterator fail-fast when the buffer is mutated after the iterator's creation.

Fix appending a `ListBuffer` to itself and subtracting a `ListBuffer` from itself.

Fix appending a `Growable` that uses the default `addAll` implementation to itself.

Fixes scala/bug#3088
Partially addresses scala/bug#12121
This PR is the beginning of a strategy to address scala/bug#12009

NthPortal commented:
I think this is definitely a community build candidate; I'm hoping this doesn't break a lot of things (if it does, that probably means a lot of bugs), but I think it's definitely something to watch out for since this will hopefully make it into a 2.13 release

sjrd commented on src/library/scala/collection/MutationTracker.scala:
```diff
@@ -0,0 +1,83 @@
+  private final class CheckedIterator[A](underlying: Iterator[A]) extends AbstractIterator[A] {
+    private[this] val checker = new Checker
+
+    def hasNext: Boolean = underlying.hasNext
```
Due to the contract of iterators, `hasNext` often actually walks the internal data structure of the collection, and is therefore also at risk of causing weird bugs like NPEs in case of concurrent access. Therefore, it should also be protected `checkMutationsForIterator()`. And yes, that's worse for performance.

martijnhoekstra replied:
That's not needed for the purpose of best-effort fast-failing though. As I understand it, the goal of this change is to prevent the user from iterating over a modified collection if we can. Emitting an error on getting the next element will accomplish that, even if hasNext throws or returns a wrong result.

NthPortal replied:
do you think we could get away with just checking in `hasNext` either? I'm guessing not...

NthPortal replied:
> That's not needed for the purpose of best-effort fast-failing though.

given that, I think checking for `hasNext` and not `next()` is probably safe enough, and safer than just checking in `next()`, since `next()` is typically only called right after `hasNext`

NthPortal commented:
I'm getting a failure with the following test, and having trouble diagnosing

> scala.tools.nsc.scaladoc.HtmlFactoryTest.(It is not a test it is a sbt.testing.SuiteSelector)

anyone know where to look?

NthPortal commented:
benchmarks:

Before
======

```
Benchmark                            (size)  Mode  Cnt         Score        Error  Units
ListBufferBenchmark.filterInPlace        10  avgt   40       209.552 ±      3.828  ns/op
ListBufferBenchmark.filterInPlace       100  avgt   40      1819.224 ±     22.649  ns/op
ListBufferBenchmark.filterInPlace      1000  avgt   40     13115.598 ±   2540.419  ns/op
ListBufferBenchmark.filterInPlace     10000  avgt   40     88563.923 ±   1244.332  ns/op
ListBufferBenchmark.flatMapInPlace1      10  avgt   40       512.107 ±      9.471  ns/op
ListBufferBenchmark.flatMapInPlace1     100  avgt   40      3106.082 ±     38.434  ns/op
ListBufferBenchmark.flatMapInPlace1    1000  avgt   40     40203.206 ±    741.022  ns/op
ListBufferBenchmark.flatMapInPlace1   10000  avgt   40    372594.646 ±  11931.681  ns/op
ListBufferBenchmark.insert               10  avgt   40       233.837 ±      2.661  ns/op
ListBufferBenchmark.insert              100  avgt   40      4695.472 ±     38.096  ns/op
ListBufferBenchmark.insert             1000  avgt   40    476238.273 ±   2797.168  ns/op
ListBufferBenchmark.insert            10000  avgt   40  48167162.543 ± 373388.643  ns/op
ListBufferBenchmark.insertAll            10  avgt   40       313.262 ±      7.922  ns/op
ListBufferBenchmark.insertAll           100  avgt   40      1435.928 ±     20.963  ns/op
ListBufferBenchmark.insertAll          1000  avgt   40     73795.740 ±    626.632  ns/op
ListBufferBenchmark.insertAll         10000  avgt   40   6567366.514 ±  29627.485  ns/op
ListBufferBenchmark.iteratorA            10  avgt   40       215.463 ±      3.137  ns/op
ListBufferBenchmark.iteratorA           100  avgt   40      1805.295 ±     50.502  ns/op
ListBufferBenchmark.iteratorA          1000  avgt   40     17333.637 ±    183.418  ns/op
ListBufferBenchmark.iteratorA         10000  avgt   40     82736.341 ±   1317.157  ns/op
ListBufferBenchmark.iteratorB            10  avgt   40       243.197 ±      4.185  ns/op
ListBufferBenchmark.iteratorB           100  avgt   40      1874.217 ±     32.496  ns/op
ListBufferBenchmark.iteratorB          1000  avgt   40      9206.308 ±    118.528  ns/op
ListBufferBenchmark.iteratorB         10000  avgt   40    142303.132 ±  27946.817  ns/op
ListBufferBenchmark.remove1              10  avgt   40       203.914 ±      3.968  ns/op
ListBufferBenchmark.remove1             100  avgt   40      1327.190 ±     23.052  ns/op
ListBufferBenchmark.remove1            1000  avgt   40    125240.886 ±   1029.013  ns/op
ListBufferBenchmark.remove1           10000  avgt   40  13557831.094 ± 137941.865  ns/op
ListBufferBenchmark.remove2              10  avgt   40       189.867 ±      4.883  ns/op
ListBufferBenchmark.remove2             100  avgt   40       754.743 ±      8.621  ns/op
ListBufferBenchmark.remove2            1000  avgt   40     40683.386 ±    620.152  ns/op
ListBufferBenchmark.remove2           10000  avgt   40   3365176.547 ±  51647.179  ns/op
ListBufferBenchmark.update               10  avgt   40       227.541 ±      1.935  ns/op
ListBufferBenchmark.update              100  avgt   40      4621.739 ±     30.248  ns/op
ListBufferBenchmark.update             1000  avgt   40    480379.841 ±   4256.451  ns/op
ListBufferBenchmark.update            10000  avgt   40  52779922.013 ± 687751.997  ns/op
```

After
=====

```
Benchmark                            (size)  Mode  Cnt         Score        Error  Units
ListBufferBenchmark.filterInPlace        10  avgt   40       196.638 ±      4.577  ns/op
ListBufferBenchmark.filterInPlace       100  avgt   40      1756.866 ±     76.259  ns/op
ListBufferBenchmark.filterInPlace      1000  avgt   40     16077.728 ±    292.931  ns/op
ListBufferBenchmark.filterInPlace     10000  avgt   40     92484.924 ±   1050.074  ns/op
ListBufferBenchmark.flatMapInPlace1      10  avgt   40       485.547 ±      8.411  ns/op
ListBufferBenchmark.flatMapInPlace1     100  avgt   40      3864.339 ±     41.703  ns/op
ListBufferBenchmark.flatMapInPlace1    1000  avgt   40     37358.350 ±    610.488  ns/op
ListBufferBenchmark.flatMapInPlace1   10000  avgt   40    320438.086 ±   4169.805  ns/op
ListBufferBenchmark.insert               10  avgt   40       208.389 ±     10.885  ns/op
ListBufferBenchmark.insert              100  avgt   40      5547.858 ±     70.386  ns/op
ListBufferBenchmark.insert             1000  avgt   40    472445.513 ±   3559.258  ns/op
ListBufferBenchmark.insert            10000  avgt   40  47403816.905 ± 409465.703  ns/op
ListBufferBenchmark.insertAll            10  avgt   40       283.284 ±      6.188  ns/op
ListBufferBenchmark.insertAll           100  avgt   40      2108.278 ±     25.333  ns/op
ListBufferBenchmark.insertAll          1000  avgt   40     73774.687 ±    750.869  ns/op
ListBufferBenchmark.insertAll         10000  avgt   40   6443110.161 ±  57711.887  ns/op
ListBufferBenchmark.iteratorA            10  avgt   40       201.963 ±      9.603  ns/op
ListBufferBenchmark.iteratorA           100  avgt   40      1563.120 ±     28.937  ns/op
ListBufferBenchmark.iteratorA          1000  avgt   40     15574.602 ±    298.189  ns/op
ListBufferBenchmark.iteratorA         10000  avgt   40     86412.182 ±   1132.029  ns/op
ListBufferBenchmark.iteratorB            10  avgt   40       225.307 ±      3.005  ns/op
ListBufferBenchmark.iteratorB           100  avgt   40      1851.259 ±     37.221  ns/op
ListBufferBenchmark.iteratorB          1000  avgt   40     16971.258 ±    285.469  ns/op
ListBufferBenchmark.iteratorB         10000  avgt   40    109750.241 ±   3137.235  ns/op
ListBufferBenchmark.remove1              10  avgt   40       172.079 ±      3.029  ns/op
ListBufferBenchmark.remove1             100  avgt   40      2122.334 ±     23.976  ns/op
ListBufferBenchmark.remove1            1000  avgt   40    127018.261 ±   1097.686  ns/op
ListBufferBenchmark.remove1           10000  avgt   40  13902997.558 ± 100696.734  ns/op
ListBufferBenchmark.remove2              10  avgt   40       164.865 ±      2.265  ns/op
ListBufferBenchmark.remove2             100  avgt   40      1489.532 ±     21.509  ns/op
ListBufferBenchmark.remove2            1000  avgt   40     40996.992 ±    452.266  ns/op
ListBufferBenchmark.remove2           10000  avgt   40   3388047.209 ±  40572.239  ns/op
ListBufferBenchmark.update               10  avgt   40       218.951 ±      8.176  ns/op
ListBufferBenchmark.update              100  avgt   40      5525.563 ±     42.342  ns/op
ListBufferBenchmark.update             1000  avgt   40    490791.195 ±   3660.844  ns/op
ListBufferBenchmark.update            10000  avgt   40  53137869.474 ± 541782.921  ns/op
```

NthPortal commented:
I'm not sure what to make of the results. neither is clearly more efficient than the other, and each have some cases with significantly better performance. I can only assume that, because my computer was doing other things in the background (even though I was not actively using it), the results are not perfectly accurate. It would certainly be valuable for someone with a dedicated benchmarkinig machine (i.e. that isn't used for anything else) to repeat the benchmarks (the addition of which is a single commit (b313b889e5cb5bb8de8caf7dfac25537a241a2b2) that can be cherry-picked to `2.13.x`). Nevertheless, my initial conclusion is that there is not a significant enough performance hit to the addition of fast-failing to the iterators to be of any concern, if there is any performance cost at all. I am happy to hear others' analysis, and to have myself disproven by further benchmarking.

Ichoran commented:
Could you add a `singletons` benchmark that just does `xs.map(x => ListBuffer(x))`?  When adding new work to a constructor, it's good to test it on small loads.  But a single object creation is a bit fast for even JMH to measure accurately.

NthPortal commented:
> Could you add a `singletons` benchmark that just does `xs.map(x => ListBuffer(x))`?

what's `xs` there? if it's a `ListBuffer`, `map` doesn't modify anything, so it doesn't log a mutation and there's not any performance impact of calling it; I wouldn't expect it to make any difference.

NthPortal commented:
> Could you add a `singletons` benchmark that just does `xs.map(x => ListBuffer(x))`?

@Ichoran what's `xs` there? if it's a `ListBuffer`, `map` doesn't modify anything, so it doesn't log a mutation and there's not any performance impact of calling it; I wouldn't expect it to make any difference.

Ichoran commented:
@NthPortal - `xs` can be a `ListBuffer`.  It doesn't really matter--but since you're set up to build `ListBuffer` anyway, why not use it?  The point is that it calls the `ListBuffer` constructor, which now has to build a `MutationChecker`.

Ichoran commented on src/library/scala/collection/mutable/ListBuffer.scala:
```diff
@@ -274,7 +288,7 @@ class ListBuffer[A]
 
   def mapInPlace(f: A => A): this.type = {
-    ensureUnaliased()
+    tracker.addMutation()
```
I don't follow why this should be `addMutation` instead of `ensureUnaliased`.

Ichoran replied:
Oops, yes I do.  Because you don't alter the old list anyway--it's the same logic as used to ensure unaliased.  Never mind!

NthPortal replied:
because it builds a completely fresh instance and then just copies references, so it doesn't matter if it's aliased. although it would be useful to set it to `aliased = false` I guess...

NthPortal commented:
> But a single object creation is a bit fast for even JMH to measure accurately.

I have no reason to believe that's correct. JMH generates a whole bunch of logic to do things that are less than the granularity of calls to `System.nanoTime()` in loops.

If you just want to benchmark the constructor, it doesn't really need sizes particularly (so it might need a separate benchmark) - I don't think there's any need to `map` it.

NthPortal commented:
```diff
- before
+ after

  Benchmark                         (size)  Mode  Cnt    Score    Error  Units
- ConstructionBenchmark.listBuffer       0  avgt   40   18.419 ±  0.293  ns/op
+ ConstructionBenchmark.listBuffer       0  avgt   40   21.373 ±  0.177  ns/op
- ConstructionBenchmark.listBuffer       1  avgt   40   25.768 ±  0.456  ns/op
+ ConstructionBenchmark.listBuffer       1  avgt   40   28.390 ±  0.273  ns/op
- ConstructionBenchmark.listBuffer      10  avgt   40  114.286 ±  1.931  ns/op
+ ConstructionBenchmark.listBuffer      10  avgt   40  113.392 ±  1.512  ns/op
- ConstructionBenchmark.listBuffer     100  avgt   40  963.965 ± 17.102  ns/op
+ ConstructionBenchmark.listBuffer     100  avgt   40  931.931 ± 17.372  ns/op
```

SethTisue commented:
> I'm getting a failure with the following test, and having trouble diagnosing

@NthPortal when I run this locally I get:

```
[error] Uncaught exception when running scala.tools.nsc.scaladoc.HtmlFactoryTest: java.util.ConcurrentModificationException: mutation occurred after iterator creation
[error] stack trace is suppressed; run last scalacheck / Test / testOnly for the full output
```

which is not a test failure per se, but an exception occurring in the test framework:

```
[error] Uncaught exception when running scala.tools.nsc.scaladoc.HtmlFactoryTest: java.util.ConcurrentModificationException: mutation occurred after iterator creation
[error] sbt.ForkMain$ForkError: java.util.ConcurrentModificationException: mutation occurred after iterator creation
[error] 	at scala.collection.MutationTracker$Checker.checkMutations(MutationTracker.scala:52)
[error] 	at scala.collection.MutationTracker$Checker.checkMutationsForIterator(MutationTracker.scala:67)
[error] 	at scala.collection.MutationTracker$CheckedIterator.hasNext(MutationTracker.scala:74)
[error] 	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
[error] 	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] 	at org.scalacheck.ScalaCheckRunner$$anon$3.execute(ScalaCheckFramework.scala:118)
[error] 	at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:304)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
...
```

after some not-very-rigorous poking around it appears to me that the `ListBuffer` in question is the one in https://github.com/typelevel/scalacheck/blob/master/src/main/scala/org/scalacheck/Properties.scala — perhaps `ScalaCheckRunner` and/or `Properties` ought to be doing some synchronization?

SethTisue commented:
> I think this is definitely a community build candidate

queued: https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/1748/

NthPortal commented:
@SethTisue it's not a synchronization/thread-safety issue; rather an issue of using an iterator after mutating the underlying collection, likely in the same thread.

Ichoran commented:
Thanks!

The difference of 2-3 ns is about what I'd expect.  Object allocations tend to take about that long.

I had forgotten what a huge penalty there was for using `from` (or the varargs companion method).

`val x = new ListBuffer; x += thingy; x` is fast, but if nobody does that, then it doesn't matter if it takes ~50% longer (3 object creations instead of 2).

I think if you write the test that way, you'll show a much bigger difference.  I don't know whether it's relevant to usage of `ListBuffer` in the wild--maybe not.

NthPortal commented:
> I don't know whether it's relevant to usage of `ListBuffer` in the wild--maybe not.

at the end of the day, people tend to use collections to put things in, so I'm not overly concerned with the empty collection case being slightly more expensive

Ichoran commented:
It's not "slightly more" if you do it with `new` and element addition.  You didn't test that.  Many lists are small, also.

But small lists tend to be created with `::` so maybe it's okay.

som-snytt commented:
`ListBuffer.empty[Int].tap(_ += 42)` just to get ichoran's goat.

NthPortal commented:
```diff
  Benchmark                         (size)  Mode  Cnt    Score    Error  Units
- ConstructionBenchmark.listBuffer       0  avgt   40    6.125 ±  0.119  ns/op
+ ConstructionBenchmark.listBuffer       0  avgt   40    8.174 ±  0.140  ns/op
- ConstructionBenchmark.listBuffer       1  avgt   40   10.803 ±  0.216  ns/op
+ ConstructionBenchmark.listBuffer       1  avgt   40   14.082 ±  0.238  ns/op
- ConstructionBenchmark.listBuffer      10  avgt   40   60.687 ±  0.634  ns/op
+ ConstructionBenchmark.listBuffer      10  avgt   40   64.039 ±  2.277  ns/op
- ConstructionBenchmark.listBuffer     100  avgt   40  576.759 ± 10.223  ns/op
+ ConstructionBenchmark.listBuffer     100  avgt   40  586.238 ±  9.666  ns/op
```

Ichoran commented:
:+1: Thank you!  I'm not sure whether this is relevant for anyone's performance, but it's good to have documented so we can think about it more precisely.

NthPortal commented:
I was going to post a benchmark showing how whether you use `new`, `from` or `to` has a much more significant effect on performance, but then I got these results:

```
Benchmark                              (size)  Mode  Cnt    Score    Error  Units
ConstructionBenchmark.listBuffer_from       0  avgt   40   17.970 ±  0.277  ns/op
ConstructionBenchmark.listBuffer_from       1  avgt   40   25.273 ±  0.368  ns/op
ConstructionBenchmark.listBuffer_from      10  avgt   40  119.745 ±  4.541  ns/op
ConstructionBenchmark.listBuffer_from     100  avgt   40  973.068 ± 12.593  ns/op
ConstructionBenchmark.listBuffer_new        0  avgt   40    6.104 ±  0.103  ns/op
ConstructionBenchmark.listBuffer_new        1  avgt   40   10.781 ±  0.223  ns/op
ConstructionBenchmark.listBuffer_new       10  avgt   40   60.370 ±  1.041  ns/op
ConstructionBenchmark.listBuffer_new      100  avgt   40  571.437 ±  9.147  ns/op
ConstructionBenchmark.listBuffer_to         0  avgt   40   14.757 ±  0.242  ns/op
ConstructionBenchmark.listBuffer_to         1  avgt   40   17.769 ±  0.235  ns/op
ConstructionBenchmark.listBuffer_to        10  avgt   40   58.702 ±  0.779  ns/op
ConstructionBenchmark.listBuffer_to       100  avgt   40  484.976 ±  7.548  ns/op
```

`to`, which itself calls `from` (after some indirection that *maybe* is optimised out by the compiler?), ought to be at least as expensive as the latter, but isn't. so... not sure what to make of that

NthPortal commented:
@SethTisue I've found the issue with that test - it's that it defines properties within another property. because properties are lazy, that the nested ones are created while the original ones are being iterated over and evaluated, which causes problems

SethTisue commented:
community build result: scopt failed (https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/1748/artifact/logs/scopt-build.log), 2 downstream projects didn't run, everything else green

NthPortal commented:
thank you for making my life easy Seth 💜

NthPortal commented:
Do we think a single project in the community build failing represents a significant fraction of the community relying on UB such that we don't want to push this forward? or, as an alternative, we could be conservative and only increment the mutation count for removals, which are always unsafe afaict?

som-snytt commented:
Uninformed Buffering?

NthPortal commented:
Undefined Behavior (popularised by C)

NthPortal commented on src/compiler/scala/tools/nsc/transform/async/AnfTransform.scala:
```diff
@@ -136,19 +136,11 @@ private[async] trait AnfTransform extends TransformUtils {
           // Definitions within stats lifted out of the `ValDef` rhs should no longer be owned by the
           // the ValDef.
-          statsIterator.foreach(_.changeOwner((currentOwner, currentOwner.owner)))
+          currentStats.iterator.foreach(_.changeOwner((currentOwner, currentOwner.owner)))
```
doesn't really need `.iterator`

NthPortal commented on test/junit/scala/collection/mutable/MutationTrackingTest.scala:
```diff
@@ -0,0 +1,142 @@
+    def runWith(c: Checker[C]): Unit
+  }
+
+  private object Suite {
```
oops, I overengineered it. but it will be highly reusable, so it's worth it

som-snytt replied:
I would like that on a t-shirt. I'm not sure which part goes on the front and which part goes on the back.

NthPortal committed:
[bug#3088][bug#12121] Fix adding/subtracting ListBuffer to/from itself

Fix appending a `ListBuffer` to itself and subtracting
a `ListBuffer` from itself.

Fix appending a `Growable` that uses the default
`addAll` implementation to itself.

NthPortal commented:
this PR is squashed, cleaned up and ready for final review.

also, I've already written the code for `ArrayBuffer` on top of this, and I will push it up if/as soon as this goes through.

Ichoran commented on test/scalacheck/scala/tools/nsc/scaladoc/HtmlFactoryTest.scala:
```diff
@@ -537,16 +537,17 @@ object HtmlFactoryTest extends Properties("HtmlFactory") {
+        ifExists(_ contains """<span class="modifier">implicit </span>""")
+      property("class:gt4s") = ifExists(_ contains """title="gt4s: $colon$colon"""")
+      property("class:gt4s of a deprecated method") =
+        ifExists(_ contains """title="gt4s: $colon$colon$colon$colon. Deprecated: """)
```
How is this related to the PR?

NthPortal replied:
so! fun story: the call to `property` appends the new property to a `ListBuffer` containing all the properties. during execution, we iterate over that `ListBuffer`, and if you create a new property during that iteration (which happens if you nest properties, which is not officially supported anyway), it breaks

Ichoran reviewed (approved):
I'm still not sold that this is the right thing to try to do (due to space and speed penalties), but the implementation looks excellent and the benchmarks are better than I feared, so if we're doing this, LGTM!

retronym commented:
@NthPortal I'm inclined to store the `mutationCount` directly as a field in the collections. This would reduce the footprint change for `ListBuffer`. It fits the `mutatationCount` into spare space in the current layout:

```
scala.collection.mutable.ListBuffer object internals:
 OFFSET  SIZE                                      TYPE DESCRIPTION                               VALUE
      0     4                                           (object header)                           01 00 00 00 (00000001 00000000 00000000 00000000) (1)
      4     4                                           (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4                                           (object header)                           49 05 02 f8 (01001001 00000101 00000010 11111000) (-134085303)
     12     4                                       int ListBuffer.mutationCount                  0
     16     4                                       int ListBuffer.len                            0
     20     1                                   boolean ListBuffer.aliased                        false
     21     3                                           (alignment/padding gap)
     24     4           scala.collection.immutable.List ListBuffer.first                          (object)
     28     4   scala.collection.immutable.$colon$colon ListBuffer.last0                          null
Instance size: 32 bytes
Space losses: 3 bytes internal + 0 bytes external = 3 bytes total

➜  pr-9174 git:(review/9174) java -Djdk.attach.allowAttachSelf=true -cp $(coursier fetch -q -p 'org.openjdk.jol:jol-cli:0.9') org.openjdk.jol.Main internals -cp $(coursier fetch -q -p 'org.scala-lang:scala-library:2.13.3') scala.collection.mutable.ListBuffer
# WARNING: Unable to attach Serviceability Agent. You can try again with escalated privileges. Two options: a) use -Djol.tryWithSudo=true to try with sudo; b) echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope
# Running 64-bit HotSpot VM.
# Using compressed oop with 3-bit shift.
# Using compressed klass with 3-bit shift.
# WARNING | Compressed references base/shifts are guessed by the experiment!
# WARNING | Therefore, computed addresses are just guesses, and ARE NOT RELIABLE.
# WARNING | Make sure to attach Serviceability Agent to get the reliable addresses.
# Objects are 8 bytes aligned.
# Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
# Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

Instantiated the sample instance via default constructor.

scala.collection.mutable.ListBuffer object internals:
 OFFSET  SIZE                                      TYPE DESCRIPTION                               VALUE
      0     4                                           (object header)                           01 00 00 00 (00000001 00000000 00000000 00000000) (1)
      4     4                                           (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4                                           (object header)                           49 05 02 f8 (01001001 00000101 00000010 11111000) (-134085303)
     12     4                                       int ListBuffer.len                            0
     16     1                                   boolean ListBuffer.aliased                        false
     17     3                                           (alignment/padding gap)
     20     4           scala.collection.immutable.List ListBuffer.first                          (object)
     24     4   scala.collection.immutable.$colon$colon ListBuffer.last0                          null
     28     4                                           (loss due to the next object alignment)
Instance size: 32 bytes
Space losses: 3 bytes internal + 4 bytes external = 7 bytes total
```

We can still share the code `CheckedIterator` code and the exception throwers.

https://github.com/scala/scala/compare/2439b992f597ce0c017176e6c4f6b34aa91a911d...retronym:review/9174?expand=1

NthPortal commented:
@retronym okay, I'll do that. arguably it's a simplification, even, since it's not obviously more complicated, and there's less code.

however, I think I'll change the function to a by-name for ease of use, and I might also put the "expected" parameters before the "actual" ones, to match common usage a la junit.

NthPortal committed:
[bug#12009] Make ListBuffer's iterator fail-fast

Make `ListBuffer`'s iterator fail-fast when the buffer is
mutated after the iterator's creation.

Co-authored-by: Jason Zaugg <jzaugg@gmail.com>

Ichoran commented:
Yes, overall I'm happier!  But, better run the benchmarks again because the by-name creates an extra object.

NthPortal commented:
@Ichoran it previously created an extra `Checker` object, so it shouldn't be all that different. I'll try to run benchmarks anyway, but they take a lot of effort as I don't have a dedicated benchmarking machine, so I don't know when I'll get to them

Ichoran commented:
@NthPortal - Before it created an extra object for every `ListBuffer`.  Now it creates an extra object every call to `.iterator()`.

som-snytt commented:
I still hope to get my eyes on this before it is merged, as I would like to be overall happier, too.

NthPortal commented:
@Ichoran what I was saying is that it previously created an extra object for every call to `.iterator` too - it created a `Checker` that the iterator wrapped

Ichoran commented:
@NthPortal - Oh, it did?  I forgot, sorry!  Then it should just be a complete win unless the indirection of the lambda prevents some important optimization.

NthPortal commented:
are we waiting on anything for this? (I'll fix the conflict shortly)

retronym commented on src/compiler/scala/tools/nsc/transform/async/AnfTransform.scala:
```diff
@@ -136,19 +136,11 @@ private[async] trait AnfTransform extends TransformUtils {
-          // But, ListBuffer.empty.iterator doesn't reflect later mutation. Luckily we can just start
-          // from the beginning of the buffer
-          if (statsIterator == null) statsIterator = currentStats.iterator
-
```
The intent of this code was to avoid `O(N^2)` cost for this transform wrt the number of statements in a block. The fail-fast iterators scupper the approach, but we can at least reduce the constant factor signficantly by avoiding the `changeOwner` for all but the new entries in the buffer:

```scala
        case ValDef(mods, name, tpt, rhs) => atOwner(tree.symbol) {
          val currentStatsLen = currentStats.length
          val expr = atOwner(currentOwner.owner)(transform(rhs))
          // Definitions within stats lifted out of the `ValDef` rhs should no longer be owned by the
          // the ValDef.
          currentStats.iterator.drop(currentStatsLen).foreach(_.changeOwner((currentOwner, currentOwner.owner)))
```

NthPortal replied:
in the previous code, the iterator was always created from the beginning `currentStats`, regardless of which expression created the iterator. since (as far as I can tell) we only ever append to `currentStats`, it still ended up iterating over all of it anyway. the comment

> Capture current cursor of a non-empty `stats` buffer so we can efficiently restrict the `changeOwner` to the newly added items...

was not, as far as I can tell, accurate in any way.

your change looks like it should work well though, and I will add it.

retronym commented on src/library/scala/collection/mutable/ListBuffer.scala:
```diff
@@ -34,13 +33,15 @@ import scala.runtime.Statics.releaseFence
   *  @define mayNotTerminateInf
   *  @define willNotTerminateInf
   */
+@SerialVersionUID(3L)
```
I think this should be explicitly set to the automatic UID generated before this PR.

retronym replied:
```
scala-launch 2.12.3 -nc -e 'println(java.io.ObjectStreamClass.lookup(classOf[scala.collection.mutable.ListBuffer[_]]).getSerialVersionUID)'
3419063961353022662
```

NthPortal replied:
I just realised this is the wrong UID, since this is for 2.**12**.3. I'll fix

retronym replied:
Oh sorry for the confusion!

retronym commented:
It's worth noting a difference between this implementation and Java's.

Some Java mutable collections provide iterators that fail fast if the underlying collection is structurally mutated (like this PR), but do offer a supported means of mutation-during-iteration by having the iterators themselves offer an API to safely mutate. See for example [`Iterator.remove`](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html#remove--) / [`ListIterator.remove`](https://docs.oracle.com/javase/8/docs/api/java/util/ListIterator.html#set-E-).

NthPortal commented:
@retronym do you have a solution to that difference - i.e. a way to iterate and mutate the collection at once?

it happens to be that `ListBuffer` can safely be appended to (and probably inserted into) during iteration, but it cannot safely be removed from. other collections (e.g. `ArrayBuffer`) cannot be safely appended to, inserted into or removed from during iteration.

it seems to me that we are in a situation where we just in general don't support that use case, and would need to design mutable iterators into the library to support it.

retronym commented:
> it seems to me that we are in a situation where we just in general don't support that use case, and would need to design mutable iterators into the library to support it.

Yep, I agree. It would have been good to have added these in 2.13.0 along with the other improvements to mutable collections like `mapInPlace`. But right now we can't add them. I still support the move to fail-fast, I just wanted to highlight this issue for the sake of discussion.

NthPortal commented:
mutating iterators sound like potentially a good candidate for scala-library-next

NthPortal commented:
@retronym thinking about it a bit more, for the equivalent of what Java gives you (i.e. just supporting removal of elements), we have `filterInPlace`. anything more than that, Java doesn't support either I don't think

som-snytt commented:
Just realized that since March I have not been sheltering but filtering in place.

SethTisue commented:
> anything more than that, Java doesn't support either I don't think

As an only-somewhat-artificial example, suppose every time there is`"foo"` in a `ListBuffer` we want to remove it _and the item after it_.

With the Java-style API with `remove` you can write:

    val it = xs.iterator()
    while (it.hasNext) {
      if (it.next() == "foo") { it.remove(); it.next(); it.remove(); }
    }

This is the sort of code CS 101 students can easily write. Whereas with the Scala-style API, solving it efficiently (especially in the absence of efficient indexed access) is a brainteaser. You can do it with `filterInPlace` by tracking state as you go in a boolean `var`, but it's a lot more difficult to write (and difficult to read) then the Java-style version.

(But if we want to continue discussing this, I'd suggest a change of venue, either a new ticket or a Discourse thread. For me, it falls in a middle zone of not-exactly-opposed-but-not-excited-about-it-either. Happy to relocate this comment in its entirety to a new thread, if one starts.)

NthPortal commented on src/library/scala/collection/mutable/ListBuffer.scala:
```diff
@@ -113,19 +120,29 @@ class ListBuffer[A]
+          ensureUnaliased()
+          val copy = ListBuffer.from(this)
+          last0.next = copy.first
+          last0 = copy.last0
```
someone forgot to update `len`, and that someone is me

NthPortal replied:
also, I never fixed `insertAll`, so it looks like I have more to do

NthPortal commented on src/library/scala/collection/mutable/ListBuffer.scala:
```diff
@@ -321,12 +358,24 @@ class ListBuffer[A]
+      val n = math.min(_replaced, _len)
+      val p = locate(i)
+      removeAfter(p, math.min(n, _len - i))
+      insertAfter(p, it)
```
this fails with a `ConcurrentModificationException` if `it` comes from `this`. might fix in a separate PR. it didn't work before this PR anyway - it failed in an unspecified way (adding/removing the wrong elements) rather than with an exception

NthPortal replied:
specifically, the following code

```scala
val b = ListBuffer(1, 2, 3)
b.patchInPlace(1, b, 1)
```

would previously result in `ListBuffer(1, 1, 3, 3)` when it should result in `ListBuffer(1, 1, 2, 3, 3)`, but now throws an exception

retronym commented on src/library/scala/collection/mutable/Growable.scala:
```diff
@@ -56,10 +56,15 @@ trait Growable[-A] extends Clearable {
-    while (it.hasNext) {
-      addOne(it.next())
+    xs match {
+      case xs: AnyRef if xs eq this => addAll(Buffer.from(xs)) // avoid mutating under our own iterator
```
nitpick: elsewhere we tend to write `if (xs.asInstanceOf[AnyRef] eq this)` which is the same from a code gen perspective. The no-op `asInstanceOf[AnyRef]` isn't pretty but it would be at least consistently so.

NthPortal replied:
done

NthPortal commented:
I think this is ready for merge?

NthPortal commented:
Travis please

som-snytt commented:
...or forever hold your peace.

som-snytt commented on src/compiler/scala/tools/nsc/transform/async/AnfTransform.scala:
```diff
@@ -136,19 +136,15 @@ private[async] trait AnfTransform extends TransformUtils {
           // Definitions within stats lifted out of the `ValDef` rhs should no longer be owned by the
           // the ValDef.
-          statsIterator.foreach(_.changeOwner((currentOwner, currentOwner.owner)))
+          currentStats.iterator.drop(oldItemsCount).foreach(_.changeOwner((currentOwner, currentOwner.owner)))
```
OK, now I understand the previous comment. Iterators concat lazily, so this works also for initial empty:
```
scala> import collection.mutable.ListBuffer
import collection.mutable.ListBuffer

scala> val xs = ListBuffer.empty[Int]
val xs: scala.collection.mutable.ListBuffer[Int] = ListBuffer()

scala> xs += 42
val res0: xs.type = ListBuffer(42)

scala> val n = xs.length
val n: Int = 1

scala> val ys = Iterator.empty[Int] ++ xs.iterator.drop(n)
val ys: Iterator[Int] = <iterator>

scala> xs += 27
val res1: xs.type = ListBuffer(42, 27)

scala> ys.next()
val res2: Int = 27
```
The `current.length` is captured by the iterator, not in a variable.

NthPortal replied:
that would also work, yes

som-snytt commented on src/library/scala/collection/mutable/Growable.scala:
```diff
@@ -56,10 +56,14 @@ trait Growable[-A] extends Clearable {
    *  @param xs   the IterableOnce producing the elements to $add.
    *  @return  the $coll itself.
    */
+  @nowarn("msg=will most likely never compare equal")
```
I have a sudden urge to tweak the error message to "`in all likelihood...`".

SethTisue replied:
"should never", perhaps

NthPortal replied:
well, except it's not "should never" - this is an example of when it should

NthPortal replied:
I feel more like it should be "will in general"

NthPortal replied:
also, this needs a category now 😜

som-snytt replied:
definitely a cat, I was just highlighting the fragility and the difficulty of ever tweaking thereafter.

som-snytt commented on src/library/scala/collection/mutable/ListBuffer.scala:
```diff
@@ -230,13 +248,29 @@ class ListBuffer[A]
-      if (idx < 0 || idx > len) throw new IndexOutOfBoundsException(s"$idx is out of bounds (min 0, max ${len-1})")
-      if (idx == len) ++=(elems)
-      else insertAfter(locate(idx), it)
+    if (idx < 0 || idx > len) throw new IndexOutOfBoundsException(s"$idx is out of bounds (min 0, max ${len-1})")
```
There was another PR about "out of bounds" which I haven't studied yet, but which lightbend course should I take?
```
scala> ListBuffer(42).insertAll(10, List(27))
java.lang.IndexOutOfBoundsException: 10 is out of bounds (min 0, max 0)
  at scala.collection.mutable.ListBuffer.insertAll(ListBuffer.scala:236)
  ... 32 elided

scala> ListBuffer(42).insertAll(1, List(27))

scala> ListBuffer.empty[Int].tap(_.insertAll(-1, List(42, 27)))
java.lang.IndexOutOfBoundsException: -1 is out of bounds (min 0, max -1)

scala>
```

NthPortal replied:
should that be `>= len`? or does passing `len` insert it at the end?

NthPortal replied:
```scala
scala> def b[U](op: ListBuffer[Int] => U): ListBuffer[Int] = {
     |   val b = ListBuffer(42)
     |   op(b)
     |   b
     | }
def b[U](op: scala.collection.mutable.ListBuffer[Int] => U): scala.collection.mutable.ListBuffer[Int]

scala> b(_.insertAll(1, List(27)))
val res2: scala.collection.mutable.ListBuffer[Int] = ListBuffer(42, 27)

scala> b(_.insertAll(0, List(27)))
val res3: scala.collection.mutable.ListBuffer[Int] = ListBuffer(27, 42)
```

som-snytt reviewed (approved):
I am too ashamed to ever attempt to write another unit test.

For the first time in a while, I miss `postfixOps` in
`(factory.newBuilder += 1 += 2 += 3 += 4).result()`
which style I am sure to emulate.

NthPortal commented:
/sync

NthPortal commented:
1 down, a lot to go
